### PR TITLE
Introduce `StreamingXContentResponse`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Similar to `ChunkedZipResponse` (#109820) this utility allows
Elasticsearch to send an `XContent`-based response constructed out of a
sequence of `ChunkedToXContent` fragments, provided in a streaming and
asynchronous fashion.

This will enable #93735 to proceed without needing to create a temporary
index to hold the intermediate results.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+                queueRefs.decRef();
+            }
+        });
+    }
+
+    private Iterator<? extends ToXContent> getChunksIterator(StreamingFragment fragment) {
+        return xContentBuilder.getRestApiVersion() == RestApiVersion.V_7
+            ? fragment.fragment().toXContentChunkedV7(params)
+            : fragment.fragment().toXContentChunked(params);
+    }
```
Comment (Abhiram98) [inline server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:130]: What do you think about moving this method to the StreamingFragment class?

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+                        assert nextReleasablesCache.isEmpty();
+
+                        final Releasable chunkStreamReleasable = () -> Releasables.closeExpectNoException(chunkStream);
+                        final var result = new ReleasableBytesReference(
+                            chunkStream.bytes(),
+                            releasables.isEmpty()
+                                ? chunkStreamReleasable
+                                : Releasables.wrap(Iterators.concat(Iterators.single(chunkStreamReleasable), releasables.iterator()))
+                        );
+                        targetStream = null;
```
Comment (Abhiram98) [inline server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:358]: How about splitting up these lines into a new method `createReleasableBytesReference`?

**Comment by Abhiram98** [comment]
Hi! I added a couple of comments about refactoring the StreamingXContentResponse class. I am happy to send you a small patch to change the code. Let me know what you think?

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:130]: Could do, but I don't really see the advantage. `StreamingFragment` doesn't have access to `xContentBuilder.getRestApiVersion()`.

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:358]: Again, I'm not seeing the benefit.

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        final var releasables = new ArrayList<Releasable>(taskCount);
+        try {
+            releasables.add(currentFragmentReleasable);
+            currentFragmentReleasable = null;
+            StreamingFragment fragment;
+            while ((fragment = fragmentQueue.poll()) != null) {
+                releasables.add(fragment.releasable());
+            }
+            assert fragmentQueue.isEmpty() : fragmentQueue.size(); // no concurrent adds
+            assert releasables.size() == taskCount - 1 || releasables.size() == taskCount - 2 : taskCount + " vs " + releasables.size();
```
Comment (nicktindall) [inline server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:232]: I'm missing something here, why would `releaseables.size() == taskCount - 2`, I get that `queueLength` is an upper bound, but this only happens after the enqueues have stopped, correct?

**Review by nicktindall — APPROVED** [review]
One minor clarification, but otherwise LGTM

File: server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java:232]: We pop one item from the queue before we start to process it, but we decrement `queueLength` when its processing is complete. So mostly `queueLength` will be counting this one extra item that isn't really in the queue any more.

**Pull request merged**